src/Microsoft.TemplateEngine.Edge/BuiltInManagedProvider/GlobalSettingsTemplatePackageProvider.cs

```diff
@@ -0,0 +1,268 @@
+                        //adding template package as non-managed
+                        list.Add(new TemplatePackage(this, entry.MountPointUri, entry.LastChangeTime));
+                    }
+                }
+                else
+                {
+                    list.Add(new TemplatePackage(this, entry.MountPointUri, entry.LastChangeTime));
+                }
+            }
+            return list;
+        }
+
+        public async Task<IReadOnlyList<CheckUpdateResult>> GetLatestVersionsAsync(IEnumerable<IManagedTemplatePackage> packages, CancellationToken cancellationToken)
+        {
+            _ = packages ?? throw new ArgumentNullException(nameof(packages));
+
+            var tasks = new List<Task<IReadOnlyList<CheckUpdateResult>>>();
+            foreach (var packagesGroupedByInstaller in packages.GroupBy(s => s.Installer))
+            {
+                tasks.Add(packagesGroupedByInstaller.Key.GetLatestVersionAsync(packagesGroupedByInstaller, this, cancellationToken));
+            }
+            await Task.WhenAll(tasks).ConfigureAwait(false);
+
+            var result = new List<CheckUpdateResult>();
+            foreach (var task in tasks)
+            {
+                result.AddRange(task.Result);
+            }
+            return result;
+        }
+
+        public async Task<IReadOnlyList<InstallResult>> InstallAsync(IEnumerable<InstallRequest> installRequests, CancellationToken cancellationToken)
+        {
+            _ = installRequests ?? throw new ArgumentNullException(nameof(installRequests));
+            if (!installRequests.Any())
+            {
+                return new List<InstallResult>();
+            }
+            using var disposable = await _globalSettings.LockAsync(cancellationToken).ConfigureAwait(false);
+            var packages = new List<TemplatePackageData>(await _globalSettings.GetInstalledTemplatePackagesAsync(cancellationToken).ConfigureAwait(false));
+            var results = await Task.WhenAll(installRequests.Select(async installRequest =>
+            {
+                var installersThatCanInstall = new List<IInstaller>();
+                foreach (var install in _installersByName.Values)
+                {
+                    if (await install.CanInstallAsync(installRequest, cancellationToken).ConfigureAwait(false))
+                    {
+                        installersThatCanInstall.Add(install);
+                    }
+                }
+                if (installersThatCanInstall.Count == 0)
+                {
+                    return InstallResult.CreateFailure(installRequest, InstallerErrorCode.UnsupportedRequest, $"{installRequest.PackageIdentifier} cannot be installed");
+                }
+
+                IInstaller installer = installersThatCanInstall[0];
+                return await InstallAsync(packages, installRequest, installer, cancellationToken).ConfigureAwait(false);
```

Comment:
These may run in parallel due to the `ConfigureAwait(false)`, is that correct? Is it thread-safe?

Reply:
Install operations (and any other installer operations) are thread safe. 
If the user specifies two exactly same requests to be handled by NuGet installer, first one to be installed wins, other will fail because packages will be downloaded with same name and installer fails if file already exists.  Folder installer doesn't care of duplicates, both will be installed however template will be loaded once. 
We may consider validation here that none of identifiers within same request are same, otherwise fail. Different requests are handled with EnsureInstallPrerequisites logic below.

Reply:
Whole code is meant to execute in parallel because of `Task.WhenAll`, and yes code is supposed to execute in parallel.

Reply:
Added better validation in https://github.com/dotnet/templating/pull/2993/commits/7ee79302ed69a4c87ef6d4ab128f96812774db2c